src/wallet/mms.cpp

```diff
@@ -0,0 +1,825 @@
+  switch(type)
+  {
+  case message_type_key_set:
+    // Result of a "prepare_multisig" command in the wallet
+    // Send the key set to all other members
+    for (size_t i = 1; i < m_members.size(); ++i) {
+      add_message(i, type, message_direction_out, content);
+    }
+    break;
+
+  case message_type_finalizing_key_set:
+    // Result of a "make_multisig" command in the wallet in case of N-1/N multisig
+    // Send the finalizing key set to all other members
+    for (size_t i = 1; i < m_members.size(); ++i) {
+      add_message(i, type, message_direction_out, content);
+    }
+    break;
+
+  case message_type_multisig_sync_data:
+    // Result of a "export_multisig_info" command in the wallet
+    // Send the sync data to all other members
+    for (size_t i = 1; i < m_members.size(); ++i) {
+      add_message(i, type, message_direction_out, content);
+    }
+    break;
+
+  case message_type_partially_signed_tx:
+    // Result of a "transfer" command in the wallet, or a "sign_multisig" command
+    // that did not yet result in the minimum number of signatures required
+    // Create a message "from me to me" as a container for the tx data
+    if (m_threshold == 1)
+    {
+      // Probably rare, but possible: The 1 signature is already enough, correct the type
+      // Easier to correct here than asking all callers to detect this rare special case
+      type = message_type_fully_signed_tx;
+    }
+    add_message(0, type, message_direction_in, content);
+    break;
+
+  case message_type_fully_signed_tx:
+    add_message(0, type, message_direction_in, content);
+    break;
+
+  default:
+    MERROR("Unknown message type " << (uint32_t)type);
+    break;
+  }
+}
+
+uint32_t message_store::add_message(uint32_t member_index, message_type type, message_direction direction,
+				    const std::string &content)
+{
+  size_t wallet_height;
+  m_wallet_state_provider->get_num_transfer_details(wallet_height);
+
+  message m;
+  m.id = m_next_message_id++;
+  m.type = type;
+  m.direction = direction;
+  m.content = content;
+  m.timestamp = time(NULL);
+  m.member_index = member_index;
+  if (direction == message_direction_out)
+  {
+    m.state = message_state_ready_to_send;
+  }
+  else {
+    m.state = message_state_waiting;
+  };
+  m.wallet_height = wallet_height;
+  m.hash = crypto::null_hash;
+  m_messages.push_back(m);
+  return m_messages.size() - 1;
+}
+
+uint32_t message_store::get_message_index_by_id(uint32_t id)
+{
+  for (size_t i = 0; i < m_messages.size(); ++i)
+  {
+    if (m_messages[i].id == id)
+    {
+      return i;
+    }
+  }
+  throw std::invalid_argument("Message id not found");
+  return 0;
+}
+
+message message_store::get_message_by_id(uint32_t id)
+{
+  return m_messages[get_message_index_by_id(id)];
+}
+
+bool message_store::any_message_of_type(message_type type, message_direction direction) const
+{
+  for (size_t i = 0; i < m_messages.size(); ++i)
```

Comment:
I would use `std::find` or `std::find_if` here. Don't think it's needed to be reimplemented